core/lib/merkle_tree/src/recovery.rs

```diff
@@ -0,0 +1,244 @@
+//! Merkle tree recovery logic.
```

Comment:
Maybe a short overview of the recovery algorithm here would be helpful for the future folks.